STM32/ACTenChannel/Core/Src/ACTenChannel.c

```diff
@@ -0,0 +1,395 @@
+        }
+
+        bsClearField(BS_OVER_TEMPERATURE_Msk);
+    }
+    else // Safety mode to avoid overheating of the board
+    {
+        stmSetGpio(fanCtrl, true);
+        
+        unsigned long now = HAL_GetTick();
+        if (now - previous > 2000)  // 2000ms. Should be aligned with the control period of heater.
+        {
+            previous = now;
+            adjustPWMDown();
+        }
+
+        bsSetError(BS_OVER_TEMPERATURE_Msk);
+    }
+
+    setBoardTemp(heatSinkTemperature);
+}
+
+/*!
+** @brief Updates the global error/status object.
+**
+** Adds the port and fan status bits into the error field, and clears the error bit if there are no
+** more errors.
+*/
+static void updateBoardStatus() 
+{
+    stmGetGpio(fanCtrl) ? bsSetField(AC_TEN_CH_PORT_x_STATUS_Msk(0)) : bsClearField(AC_TEN_CH_PORT_x_STATUS_Msk(0));
+    for(int i = 0; i < AC_TEN_CH_NUM_PORTS; i++)
+    {
+        stmGetGpio(heaterPorts[i]) ? bsSetField(AC_TEN_CH_PORT_x_STATUS_Msk(i+1)) : bsClearField(AC_TEN_CH_PORT_x_STATUS_Msk(i+1));
+    }
+
+    /* Clear the error mask if there are no error bits set any more. This logic could be done when
+    ** the (other) error bits are cleared, but doing here means it only needs to be done once */
+    bsClearError(AC_TEN_CH_No_Error_Msk);
+}
+
+/***************************************************************************************************
+** PUBLIC FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Setup function called at startup
+**
+** Checks the hardware matches this FW version, starts the USB communication and starts the ADC. 
+** Printing is synchronised with ADC, so it must be started in order to print anything over the USB
+** link
+*/
+void ACTenChannelInit(ADC_HandleTypeDef* hadc, TIM_HandleTypeDef* htim, TIM_HandleTypeDef* hDS18B20tim)
+{
+    // Always allow for DFU also if programmed on non-matching board or PCB version.
+    initCAProtocol(&caProto, usbRx);
+
+    /* Don't initialise any outputs or act on them if the board isn't correct */
+    if(boardSetup(ACTenChannel, (pcbVersion){BREAKING_MAJOR, BREAKING_MINOR}) == -1)
+    {
+        return;
+    }
+
+    static int16_t ADCBuffer[ADC_CHANNELS * ADC_CHANNEL_BUF_SIZE * 2]; // array for all ADC readings, filled by DMA.
+    ADCMonitorInit(hadc, ADCBuffer, sizeof(ADCBuffer)/sizeof(int16_t));
+    HAL_TIM_Base_Start(htim);
+    GpioInit();
+
+    DS18B20Init(hDS18B20tim, &DQ1, DQ1_GPIO_Port, DQ1_Pin);
+}
+
+/*!
+** @brief Loop function called repeatedly throughout
+** 
+** * Responds to user input
+** * Checks for ADC buffer switches (the ADC sample rate is synchronised with USB print rate - the 
+**   USB print rate should be 10 Hz, so every 400 ADC samples the buffer switches).
+** * Runs the closed loop control system for board temperature and PWMs the heaters as per user 
+**   input
+*/
+void ACTenChannelLoop(const char *bootMsg)
+{
```

Comment:
Maybe you can add the new stack trace print implemented recently on the AC board:
if(CAhandleUserInputs(&caProto, bootMsg)) {
        /* If a serious fault that required a reset occurs, print the stack trace immediately upon boot */
        if(printFaultInfo()) {
            clearFaultInfo();
            fhSaveDeposit();
        }
    };``